drv/fpga-server/src/main.rs

```diff
@@ -0,0 +1,272 @@
+    }
+}
+
+struct ServerImpl<FpgaT: Fpga> {
+    device: FpgaT,
+    device_reset_ticks: u64,
+    application: SpiDevice,
+    application_reset_ticks: u64,
+    buffer: [u8; 128],
+    decompressor: Option<gnarle::Decompressor>,
+}
+
+type RequestError = idol_runtime::RequestError<FpgaError>;
+type ReadDataLease = LenLimit<Leased<R, [u8]>, 128>;
+type WriteDataLease = LenLimit<Leased<W, [u8]>, 128>;
+
+impl<FpgaT: Fpga> idl::InOrderFpgaImpl for ServerImpl<FpgaT> {
+    fn device_enabled(&mut self, _: &RecvMessage) -> Result<u8, RequestError> {
+        Ok(self.device.device_enabled()? as u8)
+    }
+
+    fn set_device_enable(
+        &mut self,
+        _: &RecvMessage,
+        enabled: u8,
+    ) -> Result<(), RequestError> {
+        Ok(self.device.set_device_enable(enabled != 0)?)
+    }
+
+    fn reset_device(&mut self, _: &RecvMessage) -> Result<(), RequestError> {
+        Ok(self.device.reset_device(self.device_reset_ticks)?)
+    }
+
+    fn device_state(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<DeviceState, RequestError> {
+        Ok(self.device.device_state()?)
+    }
+
+    fn device_id(&mut self, _: &RecvMessage) -> Result<u32, RequestError> {
+        Ok(self.device.device_id()?)
+    }
+
+    fn application_enabled(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<u8, RequestError> {
+        Ok(self.device.application_enabled()? as u8)
+    }
+
+    fn set_application_enable(
+        &mut self,
+        _: &RecvMessage,
+        enabled: u8,
+    ) -> Result<(), RequestError> {
+        Ok(self.device.set_application_enable(enabled != 0)?)
+    }
+
+    fn reset_application(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<(), RequestError> {
+        Ok(self
+            .device
+            .reset_application(self.application_reset_ticks)?)
+    }
+
+    fn start_bitstream_load(
+        &mut self,
+        _: &RecvMessage,
+        bitstream_type: BitstreamType,
+    ) -> Result<(), RequestError> {
+        if let BitstreamType::Compressed = bitstream_type {
+            self.decompressor = Some(gnarle::Decompressor::default())
+        }
+        self.device.start_bitstream_load()?;
+        ringbuf_entry!(Trace::StartBitstreamLoad(bitstream_type));
+        Ok(())
+    }
+
+    fn continue_bitstream_load(
+        &mut self,
+        _: &RecvMessage,
+        data: LenLimit<Leased<R, [u8]>, 128>,
+    ) -> Result<(), RequestError> {
+        data.read_range(0..data.len(), &mut self.buffer[..data.len()])
+            .map_err(|_| RequestError::Fail(ClientError::WentAway))?;
+
+        let chunk = &mut &self.buffer[..data.len()];
+        let mut decompress_buffer = [0; 256];
+
+        match self.decompressor.as_mut() {
+            Some(decompressor) => {
+                while !chunk.is_empty() {
+                    let decompressed_chunk = gnarle::decompress(
+                        decompressor,
+                        chunk,
```

Comment:
It's not immediately obvious that this modifies `chunk`. Maybe don't take the reference above, so you can pass `&mut chunk` here?